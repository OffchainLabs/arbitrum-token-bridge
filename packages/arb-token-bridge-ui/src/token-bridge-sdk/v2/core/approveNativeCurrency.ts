import { EthBridger } from '@arbitrum/sdk'
import { Provider } from '@ethersproject/providers'
import { Signer } from 'ethers'

export type ApproveNativeCurrencyProps = {
  signer: Signer
  destinationChainProvider: Provider
}

export async function approveNativeCurrency({
  signer,
  destinationChainProvider
}: ApproveNativeCurrencyProps): Promise<void> {
  const ethBridger = await EthBridger.fromProvider(destinationChainProvider)

  const approveCustomFeeTokenTx = await ethBridger.approveFeeToken({
    l1Signer: signer
  })
  await approveCustomFeeTokenTx.wait()
}
