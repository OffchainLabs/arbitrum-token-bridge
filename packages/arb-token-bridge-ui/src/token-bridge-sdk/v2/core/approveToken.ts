import { Erc20Bridger } from '@arbitrum/sdk'
import { Provider } from '@ethersproject/providers'
import { Signer } from 'ethers'
import { SelectedToken } from '../BridgeTransferStarterV2'

export type ApproveTokenProps = {
  signer: Signer
  selectedToken: SelectedToken
  destinationChainProvider: Provider
}

export const approveToken = async ({
  signer,
  selectedToken,
  destinationChainProvider
}: ApproveTokenProps) => {
  const erc20Bridger = await Erc20Bridger.fromProvider(destinationChainProvider)
  const tx = await erc20Bridger.approveToken({
    erc20L1Address: selectedToken.address,
    l1Signer: signer
  })
  await tx.wait()
}
