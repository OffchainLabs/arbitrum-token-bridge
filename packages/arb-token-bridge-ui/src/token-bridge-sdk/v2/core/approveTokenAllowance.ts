import { Erc20Bridger } from '@arbitrum/sdk'
import { Provider } from '@ethersproject/providers'
import { BigNumber, Signer } from 'ethers'
import { fetchErc20Allowance } from '../../../util/TokenUtils'
import { ExternalCallback } from '../BridgeTransferStarterV2'

type ApproveTokenAllowanceProps = {
  address: string
  amount: BigNumber
  tokenAddress: string
  l1Signer: Signer
  l1Provider: Provider
  l2Provider: Provider
  spender: string
  tokenAllowanceApproval?: ExternalCallback
}

export async function approveTokenAllowance({
  address,
  amount,
  tokenAddress,
  l1Signer,
  l1Provider,
  l2Provider,
  spender,
  tokenAllowanceApproval = async () => {
    return false
  }
}: ApproveTokenAllowanceProps): Promise<boolean> {
  const allowanceForSpender = await fetchErc20Allowance({
    address: tokenAddress,
    provider: l1Provider,
    owner: address,
    spender
  })

  // need for approval - allowance exhausted
  if (!allowanceForSpender.gte(amount)) {
    const approval = await tokenAllowanceApproval()

    // if user doesn't approve
    if (!approval) return false

    // else, show ask for approval tx in wallet
    const erc20Bridger = await Erc20Bridger.fromProvider(l2Provider)
    const tx = await erc20Bridger.approveToken({
      erc20L1Address: tokenAddress,
      l1Signer
    })

    return true
  }

  // else, no need for approval
  return true
}
