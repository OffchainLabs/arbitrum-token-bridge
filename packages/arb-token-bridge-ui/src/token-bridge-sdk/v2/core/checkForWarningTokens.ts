import axios from 'axios'
import { getWarningTokenDescription } from '../../../components/TransferPanel/TransferPanelUtils'
import { DOCS_DOMAIN } from '../../../constants'

const warningTokensUrl =
  'https://raw.githubusercontent.com/OffchainLabs/arb-token-lists/aff40a59608678cfd9b034dd198011c90b65b8b6/src/WarningList/warningTokens.json'

export const checkForWarningTokens = async (
  selectedErc20ContractAddress: string
) => {
  try {
    // todo: later set this in local or session storage maybe?
    const warningTokens = (await axios.get(warningTokensUrl)).data

    const warningToken =
      selectedErc20ContractAddress &&
      warningTokens[selectedErc20ContractAddress.toLowerCase()]

    if (warningToken) {
      const description = getWarningTokenDescription(warningToken.type)
      return `${selectedErc20ContractAddress} is ${description}; it will likely have unusual behavior when deployed as as standard token to Arbitrum. It is not recommended that you deploy it. (See ${DOCS_DOMAIN}/for-devs/concepts/token-bridge/token-bridge-erc20 for more info.)`
    } else {
      // no warning. continue...
    }
  } catch (e) {
    console.warn('Failed to fetch warning tokens:', e)
  }
}
