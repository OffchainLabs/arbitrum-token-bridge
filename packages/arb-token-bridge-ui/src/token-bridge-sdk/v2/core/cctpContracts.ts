import { BigNumber, Signer, utils } from 'ethers'
import { AttestationResponse, getContracts } from '../../../hooks/CCTP/useCCTP'
import { prepareWriteContract, writeContract } from '@wagmi/core'
import { TokenMessengerAbi } from '../../../util/cctp/TokenMessengerAbi'
import { MessageTransmitterAbi } from '../../../util/cctp/MessageTransmitterAbi'
import { ERC20__factory } from '@arbitrum/sdk/dist/lib/abi/factories/ERC20__factory'

export const cctpContracts = (sourceChainId: number) => {
  const {
    tokenMessengerContractAddress,
    targetChainDomain,
    targetChainId,
    attestationApiUrl,
    usdcContractAddress,
    messageTransmitterContractAddress
  } = getContracts(sourceChainId)

  const depositForBurn = async ({
    amount,
    signer,
    recipient
  }: {
    amount: BigNumber
    signer: Signer
    recipient: string
  }) => {
    // CCTP uses 32 bytes addresses, while EVEM uses 20 bytes addresses
    const mintRecipient = utils.hexlify(
      utils.zeroPad(recipient, 32)
    ) as `0x${string}`

    const config = await prepareWriteContract({
      address: tokenMessengerContractAddress,
      abi: TokenMessengerAbi,
      functionName: 'depositForBurn',
      signer,
      args: [amount, targetChainDomain, mintRecipient, usdcContractAddress]
    })
    return writeContract(config)
  }

  const fetchAttestation = async (attestationHash: `0x${string}`) => {
    const response = await fetch(
      `${attestationApiUrl}/attestations/${attestationHash}`,
      { method: 'GET', headers: { accept: 'application/json' } }
    )

    const attestationResponse: AttestationResponse = await response.json()
    return attestationResponse
  }

  const waitForAttestation = async (attestationHash: `0x${string}`) => {
    while (true) {
      const attestation = await fetchAttestation(attestationHash)
      if (attestation.status === 'complete') {
        return attestation.attestation
      }

      await new Promise(r => setTimeout(r, 30_000))
    }
  }

  const receiveMessage = async ({
    messageBytes,
    attestation,
    signer
  }: {
    messageBytes: `0x${string}`
    attestation: `0x${string}`
    signer: Signer
  }) => {
    const config = await prepareWriteContract({
      address: messageTransmitterContractAddress,
      abi: MessageTransmitterAbi,
      functionName: 'receiveMessage',
      chainId: targetChainId,
      signer,
      args: [messageBytes, attestation]
    })
    return writeContract(config)
  }
  const approveForBurn = async (amount: BigNumber, signer: Signer) => {
    const contract = ERC20__factory.connect(usdcContractAddress, signer)
    return contract.functions.approve(tokenMessengerContractAddress, amount)
  }

  return {
    approveForBurn,
    depositForBurn,
    receiveMessage,
    fetchAttestation,
    waitForAttestation
  }
}
