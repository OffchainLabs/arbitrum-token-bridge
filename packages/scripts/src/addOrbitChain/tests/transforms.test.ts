import fs from "fs";
import path from "path";
import sharp from "sharp";
import {
  afterAll,
  afterEach,
  beforeEach,
  describe,
  expect,
  it,
  vi,
} from "vitest";
import { IncomingChainData, Issue } from "../schemas";
import {
  extractRawChainData,
  fetchAndProcessImage,
  nameToSlug,
  resizeImage,
  stripWhitespace,
  transformIncomingDataToOrbitChain,
  updateOrbitChainsFile,
} from "../transforms";
import {
  fullMockIssue,
  mockIncomingChainData,
  mockOrbitChain,
} from "./__mocks__/chainDataMocks";
import { warning } from "@actions/core";
import axios from "axios";

describe("Transforms", () => {
  describe("extractRawChainData", () => {
    it("should extract raw chain data from the issue", () => {
      expect(extractRawChainData(fullMockIssue)).toMatchSnapshot();
    });
  });

  describe("transformIncomingDataToOrbitChain", () => {
    it("should transform incoming chain data to OrbitChain format", async () => {
      const chainLogoPath = "/images/mockChain_Logo.png";
      const nativeTokenLogoPath = "/images/mockChain_NativeTokenLogo.png";

      const result = await transformIncomingDataToOrbitChain(
        mockIncomingChainData as IncomingChainData,
        chainLogoPath,
        nativeTokenLogoPath
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe("updateOrbitChainsFile", () => {
    const testData = `{
      "mainnet": [
        { "chainId": 2, "name": "Existing Chain 2" },
        { "chainId": 4, "name": "Existing Chain 4" },
        { "chainId": 1, "name": "Existing Chain 1" }
      ],
      "testnet": [
        { "chainId": 3, "name": "Existing Testnet 1" }
      ]
    }`;
    const tempFilePath = path.join(__dirname, "tempMockChains.json");

    beforeEach(() => {
      fs.writeFileSync(tempFilePath, testData, "utf8");
    });

    afterEach(() => {
      if (fs.existsSync(tempFilePath)) {
        fs.unlinkSync(tempFilePath);
      }
    });

    it("should update the orbit chains file correctly while preserving order", () => {
      const newChain = { ...mockOrbitChain, isTestnet: false, chainId: 5 };
      const result = updateOrbitChainsFile(newChain, tempFilePath);

      expect(result.mainnet.map((chain: any) => chain.chainId)).toEqual([
        2, 4, 1, 5,
      ]);
      expect(result.testnet.map((chain: any) => chain.chainId)).toEqual([3]);
      expect(result.mainnet.find((chain: any) => chain.chainId === 5)).toEqual(
        newChain
      );

      const updatedContent = fs.readFileSync(tempFilePath, "utf8");
      expect(updatedContent).toMatchSnapshot();
    });

    it("should add a new testnet chain while preserving order", () => {
      const newTestnetChain = {
        ...mockOrbitChain,
        isTestnet: true,
        chainId: 5,
      };
      const result = updateOrbitChainsFile(newTestnetChain, tempFilePath);

      expect(result.mainnet.map((chain: any) => chain.chainId)).toEqual([
        2, 4, 1,
      ]);
      expect(result.testnet.map((chain: any) => chain.chainId)).toEqual([3, 5]);
      expect(result.testnet.find((chain: any) => chain.chainId === 5)).toEqual(
        newTestnetChain
      );

      const updatedContent = fs.readFileSync(tempFilePath, "utf8");
      expect(updatedContent).toMatchSnapshot();
    });

    it("should handle updating an existing chain while preserving order", () => {
      const existingChainId = 2;
      const updatedChain = {
        ...mockOrbitChain,
        isTestnet: false,
        chainId: existingChainId,
        name: "Updated Chain",
      };
      const result = updateOrbitChainsFile(updatedChain, tempFilePath);

      expect(result.mainnet.map((chain: any) => chain.chainId)).toEqual([
        2, 4, 1,
      ]);
      expect(result.testnet.map((chain: any) => chain.chainId)).toEqual([3]);
      expect(result.mainnet.find((chain: any) => chain.chainId === 2)).toEqual(
        updatedChain
      );

      const updatedContent = fs.readFileSync(tempFilePath, "utf8");
      expect(updatedContent).toMatchSnapshot();
    });
  });

  describe("Utility Functions", () => {
    describe("stripWhitespace", () => {
      it("should remove all whitespace from a string", () => {
        expect(stripWhitespace("  Hello  World  ")).toBe("HelloWorld");
        expect(stripWhitespace("No Spaces")).toBe("NoSpaces");
        expect(stripWhitespace(" ")).toBe("");
        expect(stripWhitespace("")).toBe("");
      });
    });

    describe("nameToSlug", () => {
      it("should convert a name to a slug", () => {
        expect(nameToSlug("Hello World")).toBe("hello-world");
        expect(nameToSlug("Test Chain")).toBe("test-chain");
        expect(nameToSlug("Multiple   Spaces")).toBe("multiple-spaces");
        expect(nameToSlug("")).toBe("");
      });
    });
  });

  describe("resizeImage", () => {
    const testImagePath = path.join(__dirname, "__mocks__", "test-image.jpg");
    const resizedImagePath = path.join(
      __dirname,
      "__mocks__",
      "resized-test-image.jpg"
    );

    // COMMENT OUT TO KEEP RESIZED IMAGES
    afterEach(() => {
      // Clean up the resized image after each test
      if (fs.existsSync(resizedImagePath)) {
        fs.unlinkSync(resizedImagePath);
      }
    });

    it("should resize an image to under 100KB while maintaining aspect ratio", async () => {
      const inputBuffer = fs.readFileSync(testImagePath);
      const resizedBuffer = await resizeImage(inputBuffer);

      expect(resizedBuffer.length).toBeLessThanOrEqual(100 * 1024);

      // Save the resized image
      fs.writeFileSync(resizedImagePath, resizedBuffer);
      console.log(`Resized image saved to: ${resizedImagePath}`);

      // Additional check to ensure the file was actually saved
      expect(fs.existsSync(resizedImagePath)).toBe(true);
      const savedFileSize = fs.statSync(resizedImagePath).size;
      expect(savedFileSize).toBeLessThanOrEqual(100 * 1024);

      // Verify that the saved image can be decoded and aspect ratio is maintained
      const originalMetadata = await sharp(testImagePath).metadata();
      const resizedMetadata = await sharp(resizedImagePath).metadata();

      expect(resizedMetadata.format).toBe("jpeg");
      expect(resizedMetadata.width).toBeLessThanOrEqual(
        originalMetadata.width!
      );
      expect(resizedMetadata.height).toBeLessThanOrEqual(
        originalMetadata.height!
      );

      const originalAspectRatio =
        originalMetadata.width! / originalMetadata.height!;
      const resizedAspectRatio =
        resizedMetadata.width! / resizedMetadata.height!;
      expect(resizedAspectRatio).toBeCloseTo(originalAspectRatio, 2);
    });
  });
  describe("Image Download and Processing", () => {
    const downloadedImagePath = path.join(
      process.cwd(),
      "..",
      "..",
      "arb-token-bridge-ui",
      "public",
      "images",
      "downloaded_chain_logo.png"
    );

    // Clean up downloaded image after tests
    // Comment out the following 'after' block if you want to inspect the downloaded image
    afterAll(() => {
      if (fs.existsSync(downloadedImagePath)) {
        fs.unlinkSync(downloadedImagePath);
        console.log("Cleaned up downloaded image");
      }
    });

    it("should download, process, and save the chain logo image from fullMockIssue", async () => {
      const rawChainData = extractRawChainData(fullMockIssue);
      const imageUrl = rawChainData.chainLogo as string;

      expect(imageUrl).toBeTruthy();
      expect(imageUrl.startsWith("https://")).toBe(true);

      const { buffer, fileExtension } = await fetchAndProcessImage(imageUrl);

      expect(buffer).toBeTruthy();
      expect(buffer.length).toBeGreaterThan(0);
      expect(fileExtension).toBeTruthy();

      const fileName = "downloaded_chain_logo";
      const savedImagePath = saveImageLocally(buffer, fileName, fileExtension);

      expect(savedImagePath).toBeTruthy();
      console.log(`Image downloaded and saved to: ${savedImagePath}`);

      const fullSavePath = path.join(
        process.cwd(),
        "..",
        "..",
        "arb-token-bridge-ui",
        "public",
        savedImagePath
      );
      expect(fs.existsSync(fullSavePath)).toBe(true);

      const stats = fs.statSync(fullSavePath);
      expect(stats.size).toBeGreaterThan(0);
    });

    it("should throw an error if the image fetch fails", async () => {
      const invalidUrl = "https://example.com/nonexistent-image.png";
      await expect(fetchAndProcessImage(invalidUrl)).rejects.toThrow();
    });

    it("should correctly identify and handle SVG images without extension", async () => {
      const svgContent = `
        <svg xmlns="http://www.w3.org/2000/svg" width="100" height="100">
          <circle cx="50" cy="50" r="40" stroke="black" stroke-width="3" fill="red" />
        </svg>`;

      // Mock axios for this test
      const mockAxios = vi.spyOn(axios, "get").mockResolvedValueOnce({
        status: 200,
        data: svgContent,
        headers: {
          "content-type": "application/octet-stream", // Intentionally wrong content-type
        },
      });

      const { buffer, fileExtension } = await fetchAndProcessImage(
        "https://example.com/logo"
      );

      // Verify it detected SVG correctly
      expect(fileExtension).toBe(".svg");
      // Verify the SVG content wasn't modified
      expect(buffer.toString("utf8").trim()).toBe(svgContent.trim());

      mockAxios.mockRestore();
    });
  });

  describe("Image URL Extraction", () => {
    it("should handle both markdown and direct URL for chain and token logos", () => {
      const issue: Issue = {
        body: `
### Chain logo
![chain-logo](https://example.com/chain-logo.png)
### Native token logo
https://example.com/token-logo.png
### Chain name
Test Chain
          `,
        state: "open",
        html_url: "https://github.com/example/repo/issues/1",
      };

      const result = extractRawChainData(issue);
      expect(result.chainLogo).toBe("https://example.com/chain-logo.png");
      expect(result.nativeTokenLogo).toBe("https://example.com/token-logo.png");
    });
  });
});

const saveImageLocally = (
  imageBuffer: Buffer,
  fileName: string,
  fileExtension: string
): string => {
  const imageSavePath = `images/${fileName}${fileExtension}`;
  const fullSavePath = path.join(
    process.cwd(),
    "..",
    "..",
    "arb-token-bridge-ui",
    "public",
    imageSavePath
  );

  // Create directories if they don't exist
  const dirs = path.dirname(fullSavePath);
  if (!fs.existsSync(dirs)) {
    fs.mkdirSync(dirs, { recursive: true });
  }

  if (fs.existsSync(fullSavePath)) {
    warning(
      `${fileName} already exists at '${imageSavePath}'. Overwriting the existing image.`
    );
  }

  fs.writeFileSync(fullSavePath, imageBuffer);

  return `/${imageSavePath}`;
};
