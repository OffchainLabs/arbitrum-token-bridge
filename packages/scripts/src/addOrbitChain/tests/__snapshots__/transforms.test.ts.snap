// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Transforms > extractRawChainData > should extract raw chain data from the issue 1`] = `
{
  "chainId": "42161",
  "chainLogo": "https://github.com/user-attachments/assets/bb30cba1-242f-49a3-95bd-57e5bdee61e5",
  "childCustomGateway": "0x000000000000000000000000000000000000000E",
  "childErc20Gateway": "0x000000000000000000000000000000000000000F",
  "childGatewayRouter": "0x0000000000000000000000000000000000000010",
  "childMultiCall": "0x0000000000000000000000000000000000000011",
  "childWeth": "0x0000000000000000000000000000000000000013",
  "childWethGateway": "0x0000000000000000000000000000000000000014",
  "color": "#FF0000",
  "description": "A test chain.",
  "explorerUrl": "https://explorer.example.com",
  "name": "Test Chain",
  "nativeTokenAddress": "0x0000000000000000000000000000000000000001",
  "nativeTokenLogo": "https://github.com/user-attachments/assets/bb30cba1-242f-49a3-95bd-57e5bdee61e5",
  "nativeTokenName": "Test Token",
  "nativeTokenSymbol": "TST",
  "parentChainId": "1",
  "parentCustomGateway": "0x0000000000000000000000000000000000000007",
  "parentErc20Gateway": "0x0000000000000000000000000000000000000008",
  "parentGatewayRouter": "0x0000000000000000000000000000000000000009",
  "parentMultiCall": "0x000000000000000000000000000000000000000A",
  "parentWeth": "0x000000000000000000000000000000000000000C",
  "parentWethGateway": "0x000000000000000000000000000000000000000D",
  "rollup": "0x0000000000000000000000000000000000000005",
  "rpcUrl": "https://rpc.example.com",
}
`;

exports[`Transforms > transformIncomingDataToOrbitChain > should transform incoming chain data to OrbitChain format 1`] = `
{
  "bridgeUiConfig": {
    "color": "#FF0000",
    "nativeTokenData": {
      "decimals": 18,
      "logoUrl": "/images/mockChain_NativeTokenLogo.png",
      "name": "Test Token",
      "symbol": "TEST",
    },
    "network": {
      "description": "This is a test chain.",
      "logo": "/images/mockChain_Logo.png",
      "name": "Test Chain",
    },
  },
  "chainId": 1234567890,
  "confirmPeriodBlocks": 150,
  "ethBridge": {
    "bridge": "0x6c7FAC4edC72E86B3388B48979eF37Ecca5027e6",
    "inbox": "0x6396825803B720bc6A43c63caa1DcD7B31EB4dd0",
    "outbox": "0xc7491a559b416540427f9f112C5c98b1412c5d51",
    "rollup": "0xeedE9367Df91913ab149e828BDd6bE336df2c892",
    "sequencerInbox": "0x529a2061A1973be80D315770bA9469F3Da40D938",
  },
  "explorerUrl": "https://testexplorer.com",
  "isCustom": true,
  "isTestnet": true,
  "name": "Test Chain",
  "nativeToken": "0x0000000000000000000000000000000000000006",
  "parentChainId": 421614,
  "rpcUrl": "https://sepolia-rollup.arbitrum.io/rpc",
  "slug": "test-chain",
  "tokenBridge": {
    "childCustomGateway": "0x0000000000000000000000000000000000000014",
    "childErc20Gateway": "0x0000000000000000000000000000000000000015",
    "childGatewayRouter": "0x0000000000000000000000000000000000000016",
    "childMultiCall": "0x0000000000000000000000000000000000000017",
    "childProxyAdmin": "0x0000000000000000000000000000000000000000",
    "childWeth": "0x0000000000000000000000000000000000000019",
    "childWethGateway": "0x0000000000000000000000000000000000000020",
    "parentCustomGateway": "0x0000000000000000000000000000000000000007",
    "parentErc20Gateway": "0x0000000000000000000000000000000000000008",
    "parentGatewayRouter": "0x0000000000000000000000000000000000000009",
    "parentMultiCall": "0x0000000000000000000000000000000000000010",
    "parentProxyAdmin": "0x0000000000000000000000000000000000000000",
    "parentWeth": "0x0000000000000000000000000000000000000012",
    "parentWethGateway": "0x0000000000000000000000000000000000000013",
  },
}
`;

exports[`Transforms > updateOrbitChainsFile > should add a new testnet chain while preserving order 1`] = `
"{
  \\"mainnet\\": [
    {
      \\"chainId\\": 2,
      \\"name\\": \\"Existing Chain 2\\"
    },
    {
      \\"chainId\\": 4,
      \\"name\\": \\"Existing Chain 4\\"
    },
    {
      \\"chainId\\": 1,
      \\"name\\": \\"Existing Chain 1\\"
    }
  ],
  \\"testnet\\": [
    {
      \\"chainId\\": 3,
      \\"name\\": \\"Existing Testnet 1\\"
    },
    {
      \\"chainId\\": 5,
      \\"confirmPeriodBlocks\\": 45818,
      \\"ethBridge\\": {
        \\"bridge\\": \\"0x7dd8A76bdAeBE3BBBaCD7Aa87f1D4FDa1E60f94f\\",
        \\"inbox\\": \\"0xaE21fDA3de92dE2FDAF606233b2863782Ba046F9\\",
        \\"outbox\\": \\"0x1E400568AD4840dbE50FB32f306B842e9ddeF726\\",
        \\"rollup\\": \\"0xC47DacFbAa80Bd9D8112F4e8069482c2A3221336\\",
        \\"sequencerInbox\\": \\"0x995a9d3ca121D48d21087eDE20bc8acb2398c8B1\\"
      },
      \\"nativeToken\\": \\"0x4Cb9a7AE498CEDcBb5EAe9f25736aE7d428C9D66\\",
      \\"explorerUrl\\": \\"https://explorer.xai-chain.net\\",
      \\"rpcUrl\\": \\"https://xai-chain.net/rpc\\",
      \\"isCustom\\": true,
      \\"isTestnet\\": true,
      \\"name\\": \\"Xai\\",
      \\"slug\\": \\"xai\\",
      \\"parentChainId\\": 42161,
      \\"tokenBridge\\": {
        \\"parentGatewayRouter\\": \\"0x22CCA5Dc96a4Ac1EC32c9c7C5ad4D66254a24C35\\",
        \\"childGatewayRouter\\": \\"0xd096e8dE90D34de758B0E0bA4a796eA2e1e272cF\\",
        \\"parentErc20Gateway\\": \\"0xb591cE747CF19cF30e11d656EB94134F523A9e77\\",
        \\"childErc20Gateway\\": \\"0x0c71417917D24F4A6A6A55559B98c5cCEcb33F7a\\",
        \\"parentCustomGateway\\": \\"0xb15A0826d65bE4c2fDd961b72636168ee70Af030\\",
        \\"childCustomGateway\\": \\"0x96551194230725c72ACF8E9573B1382CCBC70635\\",
        \\"parentWethGateway\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childWethGateway\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentWeth\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childWeth\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentProxyAdmin\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childProxyAdmin\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentMultiCall\\": \\"0x90B02D9F861017844F30dFbdF725b6aa84E63822\\",
        \\"childMultiCall\\": \\"0xEEC168551A85911Ec3A905e0561b656979f3ea67\\"
      },
      \\"bridgeUiConfig\\": {
        \\"color\\": \\"#F30019\\",
        \\"network\\": {
          \\"name\\": \\"Xai\\",
          \\"logo\\": \\"/images/XaiLogo.svg\\",
          \\"description\\": \\"A chain for Web2 and Web3 gamers to play blockchain games.\\"
        },
        \\"nativeTokenData\\": {
          \\"name\\": \\"Xai\\",
          \\"symbol\\": \\"XAI\\",
          \\"decimals\\": 18,
          \\"logoUrl\\": \\"/images/XaiLogo.svg\\"
        }
      }
    }
  ]
}"
`;

exports[`Transforms > updateOrbitChainsFile > should handle updating an existing chain while preserving order 1`] = `
"{
  \\"mainnet\\": [
    {
      \\"chainId\\": 2,
      \\"confirmPeriodBlocks\\": 45818,
      \\"ethBridge\\": {
        \\"bridge\\": \\"0x7dd8A76bdAeBE3BBBaCD7Aa87f1D4FDa1E60f94f\\",
        \\"inbox\\": \\"0xaE21fDA3de92dE2FDAF606233b2863782Ba046F9\\",
        \\"outbox\\": \\"0x1E400568AD4840dbE50FB32f306B842e9ddeF726\\",
        \\"rollup\\": \\"0xC47DacFbAa80Bd9D8112F4e8069482c2A3221336\\",
        \\"sequencerInbox\\": \\"0x995a9d3ca121D48d21087eDE20bc8acb2398c8B1\\"
      },
      \\"nativeToken\\": \\"0x4Cb9a7AE498CEDcBb5EAe9f25736aE7d428C9D66\\",
      \\"explorerUrl\\": \\"https://explorer.xai-chain.net\\",
      \\"rpcUrl\\": \\"https://xai-chain.net/rpc\\",
      \\"isCustom\\": true,
      \\"isTestnet\\": false,
      \\"name\\": \\"Updated Chain\\",
      \\"slug\\": \\"xai\\",
      \\"parentChainId\\": 42161,
      \\"tokenBridge\\": {
        \\"parentGatewayRouter\\": \\"0x22CCA5Dc96a4Ac1EC32c9c7C5ad4D66254a24C35\\",
        \\"childGatewayRouter\\": \\"0xd096e8dE90D34de758B0E0bA4a796eA2e1e272cF\\",
        \\"parentErc20Gateway\\": \\"0xb591cE747CF19cF30e11d656EB94134F523A9e77\\",
        \\"childErc20Gateway\\": \\"0x0c71417917D24F4A6A6A55559B98c5cCEcb33F7a\\",
        \\"parentCustomGateway\\": \\"0xb15A0826d65bE4c2fDd961b72636168ee70Af030\\",
        \\"childCustomGateway\\": \\"0x96551194230725c72ACF8E9573B1382CCBC70635\\",
        \\"parentWethGateway\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childWethGateway\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentWeth\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childWeth\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentProxyAdmin\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childProxyAdmin\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentMultiCall\\": \\"0x90B02D9F861017844F30dFbdF725b6aa84E63822\\",
        \\"childMultiCall\\": \\"0xEEC168551A85911Ec3A905e0561b656979f3ea67\\"
      },
      \\"bridgeUiConfig\\": {
        \\"color\\": \\"#F30019\\",
        \\"network\\": {
          \\"name\\": \\"Xai\\",
          \\"logo\\": \\"/images/XaiLogo.svg\\",
          \\"description\\": \\"A chain for Web2 and Web3 gamers to play blockchain games.\\"
        },
        \\"nativeTokenData\\": {
          \\"name\\": \\"Xai\\",
          \\"symbol\\": \\"XAI\\",
          \\"decimals\\": 18,
          \\"logoUrl\\": \\"/images/XaiLogo.svg\\"
        }
      }
    },
    {
      \\"chainId\\": 4,
      \\"name\\": \\"Existing Chain 4\\"
    },
    {
      \\"chainId\\": 1,
      \\"name\\": \\"Existing Chain 1\\"
    }
  ],
  \\"testnet\\": [
    {
      \\"chainId\\": 3,
      \\"name\\": \\"Existing Testnet 1\\"
    }
  ]
}"
`;

exports[`Transforms > updateOrbitChainsFile > should update the orbit chains file correctly while preserving order 1`] = `
"{
  \\"mainnet\\": [
    {
      \\"chainId\\": 2,
      \\"name\\": \\"Existing Chain 2\\"
    },
    {
      \\"chainId\\": 4,
      \\"name\\": \\"Existing Chain 4\\"
    },
    {
      \\"chainId\\": 1,
      \\"name\\": \\"Existing Chain 1\\"
    },
    {
      \\"chainId\\": 5,
      \\"confirmPeriodBlocks\\": 45818,
      \\"ethBridge\\": {
        \\"bridge\\": \\"0x7dd8A76bdAeBE3BBBaCD7Aa87f1D4FDa1E60f94f\\",
        \\"inbox\\": \\"0xaE21fDA3de92dE2FDAF606233b2863782Ba046F9\\",
        \\"outbox\\": \\"0x1E400568AD4840dbE50FB32f306B842e9ddeF726\\",
        \\"rollup\\": \\"0xC47DacFbAa80Bd9D8112F4e8069482c2A3221336\\",
        \\"sequencerInbox\\": \\"0x995a9d3ca121D48d21087eDE20bc8acb2398c8B1\\"
      },
      \\"nativeToken\\": \\"0x4Cb9a7AE498CEDcBb5EAe9f25736aE7d428C9D66\\",
      \\"explorerUrl\\": \\"https://explorer.xai-chain.net\\",
      \\"rpcUrl\\": \\"https://xai-chain.net/rpc\\",
      \\"isCustom\\": true,
      \\"isTestnet\\": false,
      \\"name\\": \\"Xai\\",
      \\"slug\\": \\"xai\\",
      \\"parentChainId\\": 42161,
      \\"tokenBridge\\": {
        \\"parentGatewayRouter\\": \\"0x22CCA5Dc96a4Ac1EC32c9c7C5ad4D66254a24C35\\",
        \\"childGatewayRouter\\": \\"0xd096e8dE90D34de758B0E0bA4a796eA2e1e272cF\\",
        \\"parentErc20Gateway\\": \\"0xb591cE747CF19cF30e11d656EB94134F523A9e77\\",
        \\"childErc20Gateway\\": \\"0x0c71417917D24F4A6A6A55559B98c5cCEcb33F7a\\",
        \\"parentCustomGateway\\": \\"0xb15A0826d65bE4c2fDd961b72636168ee70Af030\\",
        \\"childCustomGateway\\": \\"0x96551194230725c72ACF8E9573B1382CCBC70635\\",
        \\"parentWethGateway\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childWethGateway\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentWeth\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childWeth\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentProxyAdmin\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"childProxyAdmin\\": \\"0x0000000000000000000000000000000000000000\\",
        \\"parentMultiCall\\": \\"0x90B02D9F861017844F30dFbdF725b6aa84E63822\\",
        \\"childMultiCall\\": \\"0xEEC168551A85911Ec3A905e0561b656979f3ea67\\"
      },
      \\"bridgeUiConfig\\": {
        \\"color\\": \\"#F30019\\",
        \\"network\\": {
          \\"name\\": \\"Xai\\",
          \\"logo\\": \\"/images/XaiLogo.svg\\",
          \\"description\\": \\"A chain for Web2 and Web3 gamers to play blockchain games.\\"
        },
        \\"nativeTokenData\\": {
          \\"name\\": \\"Xai\\",
          \\"symbol\\": \\"XAI\\",
          \\"decimals\\": 18,
          \\"logoUrl\\": \\"/images/XaiLogo.svg\\"
        }
      }
    }
  ],
  \\"testnet\\": [
    {
      \\"chainId\\": 3,
      \\"name\\": \\"Existing Testnet 1\\"
    }
  ]
}"
`;
